File: sources/unity/Assets/Plugins/General.Typescript/core/function/Function.cs
using System;

namespace General.Typescript
{
    public delegate IntPtr StaticFunctionCallback(IntPtr handle, IntPtr arguments, int count);
    public delegate IntPtr InstanceFunctionCallback(IntPtr caller, IntPtr handle, string name, IntPtr arguments, int count);

    public abstract class Function : Base
    {
        internal Type DeclaredType { get { return (mParent as Class)?.Type ?? null; } }

        internal Function(IntPtr context, IntPtr handle, string name, Base parent) : base(context, handle, name, parent) { }
    }

    public abstract class StaticFunctionBase : Function
    {
        internal StaticFunctionBase(IntPtr context, IntPtr handle, string name, Base parent) : base(context, handle, name, parent) { }

        internal abstract IntPtr Invoke(IntPtr parameters, int count);
    }

    public class StaticFunction<TValue> : StaticFunctionBase
    {
        protected Func<Type, string, Parameters, TValue> mCallback = null;

        internal StaticFunction(IntPtr context, IntPtr handle, string name, Base parent, Func<Type, string, Parameters, TValue> callback) : base(context, handle, name, parent)
        {
            mCallback = callback;
        }

        internal override IntPtr Invoke(IntPtr parameters, int count)
        {
            if (null == mCallback)
            {
                return IntPtr.Zero;
            }
            TValue result = mCallback.Invoke(this.DeclaredType, mName, new Parameters(parameters, count));
            return null == result ? IntPtr.Zero : Entry.ReturnResultToJavascript(result);
        }
    }

    public class StaticFunction : StaticFunctionBase
    {
        protected Action<Type, string, Parameters> mCallback = null;

        internal StaticFunction(IntPtr context, IntPtr handle, string name, Base parent, Action<Type, string, Parameters> callback) : base(context, handle, name, parent)
        {
            mCallback = callback;
        }

        internal override IntPtr Invoke(IntPtr parameters, int count)
        {
            if (null != mCallback)
            {
                mCallback.Invoke((mParent as Class)?.Type ?? null, mName, new Parameters(parameters, count));
            }
            return IntPtr.Zero;
        }
    }

    public abstract class InstanceFunctionBase : Function
    {
        internal InstanceFunctionBase(IntPtr context, IntPtr handle, string name, Base parent) : base(context, handle, name, parent) { }

        internal abstract IntPtr Invoke(object owner, IntPtr parameters, int count);

        internal void Bind(IntPtr target)
        {
            //Entry.General_Typescript_Class_BindInstanceFunction(mContext, target, mName);
        }
    }

    public class InstanceFunction<TOwner> : InstanceFunctionBase
    {
        protected Action<TOwner, string, Parameters> mCallback = null;

        internal InstanceFunction(IntPtr context, IntPtr handle, string name, Base parent, Action<TOwner, string, Parameters> callback) : base(context, handle, name, parent)
        {
            mCallback = callback;
        }

        internal override IntPtr Invoke(object owner, IntPtr parameters, int count)
        {
            mCallback?.Invoke((TOwner)owner, mName, new Parameters(parameters, count));
            return IntPtr.Zero;
        }
    }

    public class InstanceFunction<TOwner, TValue> : InstanceFunctionBase
    {
        protected Func<TOwner, string, Parameters, TValue> mCallback = null;

        internal InstanceFunction(IntPtr context, IntPtr handle, string name, Base parent, Func<TOwner, string, Parameters, TValue> callback) : base(context, handle, name, parent)
        {
            mCallback = callback;
        }

        internal override IntPtr Invoke(object owner, IntPtr parameters, int count)
        {
            if (null == mCallback)
            {
                return IntPtr.Zero;
            }
            TValue result = mCallback.Invoke((TOwner)owner, mName, new Parameters(parameters, count));
            return Entry.ReturnResultToJavascript(result);
        }
    }
}

File: sources/platforms/typescript/core/function/Function.cpp
#include "Function.h"
#include "delegates.h"
#include "core/function/Parameters.h"
#include "../object/Object.h"
#include "../class/Class.h"

#ifdef __APPLE__
extern JSValueRef onStaticFunctionCall (JSContextRef context, JSObjectRef function, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception);
#else
extern void onStaticFunctionCall(const FunctionCallbackInfo<Value>& info);
#endif

ASSIGN_BASE_CHILD_TYPE(TypescriptFunction, BaseChildTypes::BaseChildType_Function);
ASSIGN_BASE_CHILD_TYPE(TypescriptStaticFunction, BaseChildTypes::BaseChildType_StaticFunction);
ASSIGN_BASE_CHILD_TYPE(TypescriptInstanceFunction, BaseChildTypes::BaseChildType_InstanceFunction);

TypescriptFunction::TypescriptFunction(const BaseChildTypes& type, const std::string& name, Base* parent, JS_FUNCTION_CALLBACK_TYPE callback) : Base(TYPE | type, name, parent), mCallback(callback) { }

TypescriptFunction::~TypescriptFunction() { }

void TypescriptFunction::bind()
{
	if (mJsObject) return;
	mJsObject = JsObject::Create(INTERNAL_JS_CONTEXT_NAME, this, mCallback, this);
    mJsObject->Set(NAME, mName);
    mJsObject->Set(FULLNAME, mFullname);
    mJsObject->MakePersistent();
}

void TypescriptFunction::BindTo(JsObject* target)
{
	if (!mJsObject)
	{
		this->Base::Bind();
	}
	target->Set(mName, mJsObject);
}

TypescriptStaticFunction* TypescriptStaticFunction::Create(const std::string& name, Base* parent, JS_FUNCTION_CALLBACK_TYPE callback)
{
	return Base::Create<TypescriptStaticFunction>(name, parent, VALID_DEFAULT(callback, onStaticFunctionCall));
}

TypescriptStaticFunction::TypescriptStaticFunction(const std::string& name, Base* parent, JS_FUNCTION_CALLBACK_TYPE callback) : TypescriptFunction(TYPE, name, parent, callback) { }

TypescriptStaticFunction::~TypescriptStaticFunction() { }

extern StaticFunctionCallback sStaticFunctionCallback;
extern DEFINE_JS_FUNCTION_CALLBACK(onInstanceFunctionCall);

JsObject* TypescriptStaticFunction::Invoke(JS_PARAMETERS)
{
#if USE_ENGINE_V8
    Parameters parameters(info);
#else
    Parameters parameters(arguments, argumentCount);
#endif
	JsObject* result = sStaticFunctionCallback(this, parameters.address(), parameters.count());
	return result;
}

TypescriptInstanceFunction* TypescriptInstanceFunction::Create(const std::string& name, Base* parent, JS_FUNCTION_CALLBACK_TYPE callback)
{
	return Base::Create<TypescriptInstanceFunction>(name, parent, VALID_DEFAULT(callback, onInstanceFunctionCall));
}

TypescriptInstanceFunction::TypescriptInstanceFunction(const std::string& name, Base* parent, JS_FUNCTION_CALLBACK_TYPE callback) : TypescriptFunction(TYPE, name, parent, callback) { }

TypescriptInstanceFunction::~TypescriptInstanceFunction() { }

void TypescriptInstanceFunction::Bind()
{
	assert(mParent);

	this->bind();
	this->bindNativeInstance();
}

JsObject* TypescriptInstanceFunction::Invoke(JS_PARAMETERS)
{
#if DEBUG
	assert(0);
#endif
	return nullptr;
}
